> builds for humble (not for foxy) features tested
> 
> * prefix
> * compresed
> * binary
> * prefix
> * fixed_frame
> 
> `ros2 run pcl_ros pointcloud_to_pcd --ros-args -r input:=os_node/points -p prefix:=abc_ -p binary:=true -p compressed:=true -p fixed_frame:=base_link`
> 
> Everything works except fixed_frame feature because my (ouster) lidar frame is on tf_static i get the following error.
> 
> ```
> [WARN] [1712221819.952296406] [transform_listener_impl_556e0f932ee0]: New publisher discovered on topic '/tf_static', offering incompatible QoS. No messages will be sent to it. Last incompatible policy: DURABILITY_QOS_POLICY
> [ERROR] [1712221820.239036841] [pointcloud_to_pcd]: Transform error: "os_sensor" passed to lookupTransform argument source_frame does not exist. 
> ```

Thank you for your testing.

fixed_frame is an option that can be used when the sensor is connected to another tf.
For example, the following command broadcasts a relative position with x=2, yaw=90deg.

`ros2 run tf2_ros static_transform_publisher 2 0 0 0 0 0.707 0.707 base_link velodyne`

I would consider implementing the following: if tf cannot be obtained, a warning is displayed and the fixed_frame option is ignored. <- done

--- a/pcl_ros/tools/pointcloud_to_pcd.cpp
+++ b/pcl_ros/tools/pointcloud_to_pcd.cpp
@@ -65,6 +65,7 @@ private:
   std::string prefix_;
   bool binary_;
   bool compressed_;
+  bool rgb_;
   std::string fixed_frame_;
   tf2_ros::Buffer tf_buffer_;
   tf2_ros::TransformListener tf_listener_;
@@ -114,48 +115,70 @@ public:
       q = transform.rotation().cast<float>();
     }
 
-    pcl::PointCloud<pcl::PointXYZRGB> cloud;
-    pcl::fromROSMsg(*cloud_msg, cloud);
-    for (size_t i = 0; i < cloud.size(); ++i) {
-      Eigen::Vector3f pt = cloud[i].getVector3fMap();
-      pt = q * pt + v.head<3>();
-      cloud[i].x = pt[0];
-      cloud[i].y = pt[1];
-      cloud[i].z = pt[2];
-    }
-
     std::stringstream ss;
     ss << prefix_ << cloud_msg->header.stamp.sec << "." << cloud_msg->header.stamp.nanosec <<
       ".pcd";
     RCLCPP_INFO(this->get_logger(), "Writing to %s", ss.str().c_str());
+    if (rgb_) {
+      pcl::PointCloud<pcl::PointXYZRGB> cloud;
+      pcl::fromROSMsg(*cloud_msg, cloud);
+      transformPointCloud(cloud, cloud, v, q);
+      writePCDFile(ss.str(), cloud);
+    } else {
+      pcl::PointCloud<pcl::PointXYZ> cloud;
+      pcl::fromROSMsg(*cloud_msg, cloud);
+      transformPointCloud(cloud, cloud, v, q);
+      writePCDFile(ss.str(), cloud);
+    }
+  }
 
+  template<typename T>
+  void transformPointCloud(
+    const pcl::PointCloud<T> & cloud_in, pcl::PointCloud<T> & cloud_out,
+    const Eigen::Vector4f & v, const Eigen::Quaternionf & q)
+  {
+    cloud_out = cloud_in;
+    for (size_t i = 0; i < cloud_in.size(); ++i) {
+      Eigen::Vector3f pt = cloud_in[i].getVector3fMap();
+      pt = q * pt + v.head<3>();
+      cloud_out[i].x = pt[0];
+      cloud_out[i].y = pt[1];
+      cloud_out[i].z = pt[2];
+    }
+  }
+
+  template<typename T>
+  void writePCDFile(const std::string & filename, const pcl::PointCloud<T> & cloud)
+  {
     pcl::PCDWriter writer;
     if (binary_) {
       if (compressed_) {
-        writer.writeBinaryCompressed(ss.str(), cloud);
+        writer.writeBinaryCompressed(filename, cloud);
       } else {
-        writer.writeBinary(ss.str(), cloud);
+        writer.writeBinary(filename, cloud);
       }
     } else {
-      writer.writeASCII(ss.str(), cloud, 8);
+      writer.writeASCII(filename, cloud, 8);
     }
   }
 
   ////////////////////////////////////////////////////////////////////////////////
   explicit PointCloudToPCD(const rclcpp::NodeOptions & options)
   : rclcpp::Node("pointcloud_to_pcd", options),
-    binary_(false), compressed_(false),
+    binary_(false), compressed_(false), rgb_(false),
     tf_buffer_(this->get_clock()), tf_listener_(tf_buffer_)
   {
     this->declare_parameter("prefix", prefix_);
     this->declare_parameter("fixed_frame", fixed_frame_);
     this->declare_parameter("binary", binary_);
     this->declare_parameter("compressed", compressed_);
+    this->declare_parameter("rgb", rgb_);
 
     this->get_parameter("prefix", prefix_);
     this->get_parameter("fixed_frame", fixed_frame_);
     this->get_parameter("binary", binary_);
     this->get_parameter("compressed", compressed_);
+    this->get_parameter("rgb", rgb_);
 
     auto sensor_qos = rclcpp::SensorDataQoS();
     sub_ = this->create_subscription<sensor_msgs::msg::PointCloud2>(

--- a/pcl_ros/tools/pointcloud_to_pcd.cpp
+++ b/pcl_ros/tools/pointcloud_to_pcd.cpp
@@ -66,6 +66,7 @@ private:
   bool binary_;
   bool compressed_;
   bool rgb_;
+  bool use_transform_;
   std::string fixed_frame_;
   tf2_ros::Buffer tf_buffer_;
   tf2_ros::TransformListener tf_listener_;
@@ -83,6 +84,7 @@ public:
     Eigen::Vector4f v = Eigen::Vector4f::Zero();
     Eigen::Quaternionf q = Eigen::Quaternionf::Identity();
     if (!fixed_frame_.empty()) {
+      use_transform_ = false;
       try {
         geometry_msgs::msg::TransformStamped transform;
         transform = tf_buffer_.lookupTransform(
@@ -99,20 +101,24 @@ public:
           transform.transform.rotation.x,
           transform.transform.rotation.y,
           transform.transform.rotation.z);
-      } catch (tf2::TransformException & ex) {
-        RCLCPP_ERROR(this->get_logger(), "Transform error: %s", ex.what());
-        return;
-      }
 
-      Eigen::Affine3d transform;
-      transform =
-        tf2::transformToEigen(
-        tf_buffer_.lookupTransform(
-          fixed_frame_, cloud_msg->header.frame_id,
-          cloud_msg->header.stamp));
-      v = Eigen::Vector4f::Zero();
-      v.head<3>() = transform.translation().cast<float>();
-      q = transform.rotation().cast<float>();
+        Eigen::Affine3d transform_eigen;
+        transform_eigen =
+          tf2::transformToEigen(
+          tf_buffer_.lookupTransform(
+            fixed_frame_, cloud_msg->header.frame_id,
+            cloud_msg->header.stamp));
+        v = Eigen::Vector4f::Zero();
+        v.head<3>() = transform_eigen.translation().cast<float>();
+        q = transform_eigen.rotation().cast<float>();
+        use_transform_ = true;
+      } catch (tf2::LookupException & ex) {
+        RCLCPP_WARN(this->get_logger(), "skip transform: %s", ex.what());
+      } catch (tf2::TransformException & ex) {
+        RCLCPP_ERROR(this->get_logger(), "skip transform: %s", ex.what());
+      }
+    } else {
+      use_transform_ = false;
     }
 
     std::stringstream ss;
@@ -122,12 +128,16 @@ public:
     if (rgb_) {
       pcl::PointCloud<pcl::PointXYZRGB> cloud;
       pcl::fromROSMsg(*cloud_msg, cloud);
-      transformPointCloud(cloud, cloud, v, q);
+      if (use_transform_) {
+        transformPointCloud(cloud, cloud, v, q);
+      }
       writePCDFile(ss.str(), cloud);
     } else {
       pcl::PointCloud<pcl::PointXYZ> cloud;
       pcl::fromROSMsg(*cloud_msg, cloud);
-      transformPointCloud(cloud, cloud, v, q);
+      if (use_transform_) {
+        transformPointCloud(cloud, cloud, v, q);
+      }
       writePCDFile(ss.str(), cloud);
     }
   }

